### PR TITLE
Fixed compiler tests related to generics / substution map and some other misc changes

### PR DESCRIPTION
**Pull request by mhong**
There are still test failures related to string attrs, which will be fixed in a later PR.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/TFDevicePartition.cpp [hunk 1]
```diff
@@ -798,11 +811,22 @@ class DevicePartitionerImpl
...
-          builtinInst->getSubstitutions(), operands);
+      // The operand must have been produced by a `builtin` inst or an
+      // UncheckedRefCastInst.
+      SubstitutionMap subMap;
+      if (auto *builtinInst = dyn_cast<BuiltinInst>(operandInst)) {
+        subMap = builtinInst->getSubstitutions();
+      } else {
+          assert(isa<UncheckedRefCastInst>(operandInst));
+          auto tensorTy = operandInst->getResults()[0]->getType();
+          assert(isTensorFlowValue(tensorTy));
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/TFDevicePartition.cpp]: Indentation is wrong here

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 2]
```diff
@@ -80,22 +80,15 @@ static CanType getSingleElementDeclFieldType(NominalTypeDecl *decl) {
...
-  builder.addGenericParameter(GenericTypeParamType::get(0, 0, ctx));
-  auto *genSig = std::move(builder).computeGenericSignature(SourceLoc());
-  return SubstitutionMap::get(genSig, [&](SubstitutableType *t) { return ty; },
-                              MakeAbstractConformanceForGenericType());
+static SubstitutionMap
+getSingleSubstitutionMapForFunction(SILFunction *f, Type ty,
+                                    SILModule &userModule) {
+  auto *loadedFunc = lookupOrLinkFunction(f->getName(), userModule);
+  assert(loadedFunc->getGenericEnvironment());
+  auto *genericSig = loadedFunc->getGenericEnvironment()->getGenericSignature();
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: Assert that genericSig->getGenericParams().size() is 1?

File: lib/SILOptimizer/Mandatory/TFPartition.cpp [hunk 3]
```diff
@@ -80,22 +80,15 @@ static CanType getSingleElementDeclFieldType(NominalTypeDecl *decl) {
...
+  return SubstitutionMap::get(genericSig,
+                              [&](SubstitutableType *t) { return ty; },
+                              LookUpConformanceInSignature(*genericSig));
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/TFPartition.cpp]: I've seen these three lines repeated elsewhere. Maybe factor it out into a new utility method?

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp [hunk 4]
```diff
@@ -257,6 +260,16 @@ SILFunction *tf::findSILFunctionForRequiredProtocolMember(
   return nullptr;
 }
 
+SubstitutionMap tf::getSingleSubstitutionMapForElementType(Type ty,
+                                                           ASTContext &ctx) {
+  auto builder = GenericSignatureBuilder(ctx);
```
Comment (slavapestov) [inline lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Does it really make sense to create a new builder every time? Why don't you store this trivial signature somewhere?

File: lib/SILOptimizer/Mandatory/TFDevicePartition.cpp (hunk 1 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFDevicePartition.cpp]: Fixed.

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 2 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFPartition.cpp (hunk 3 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFPartition.cpp]: Created tf::getSingleSubstitutionMapForElementTypeAndSignature() in TFUtilities.cpp.

File: lib/SILOptimizer/Mandatory/TFUtilities.cpp (hunk 4 above)
Comment (mhong) [reply lib/SILOptimizer/Mandatory/TFUtilities.cpp]: Added a FIXME comment -- favoring code simplicity for now, and will address it if there is non-trivial runtime overhead.

**Comment by mhong** [comment]
Thank you, @slavapestov for the great comments!

**Comment by mhong** [comment]
@lattner @rxwei PTAL -- is this PR good to go?

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
Sorry for the delayed review. LGTM.

**Review by lattner — APPROVED** [review]
Nice!

**Comment by mhong** [comment]
Thanks!

**Pull request merged**